[RFC] another python 3 port

geier opened the pull request:
This python 3 port should make alot python 3.3+ compatible.

I will now start to use alot with python 3 and fix any issues I come across, but please do leave comments as well (or give it a go under python 3).

geier commented:
@pazz would you mind six as a new dependency? That would make the imports and defining string and byte_types much nicer looking.

pazz commented:
I'm not sure how much trouble this will be for packagers.
Can anyone comment on the availability of this package
and if it is advisable to use it?

pazz commented:
regarding `six`: I think I'd rather not, if all it brings is some cosmetic improvements to the code.

geier commented:
(I posted this before as a comment under a commit that I squashed)

this currently still breaks at a rather importing step: sending mails.

First twisted.internet.FileDescriptor.write() really does not want unicode, but than msmtp fails with

alot.account.SendingMailFailed: msmtp -a lostpackets -t failed with code None:
A process has ended with a probable error condition: process ended by signal 13.

I'll look closer into it next week.

lucc commented:
I think you missed some print statements in a doctest: https://github.com/pazz/alot/blob/master/alot/helper.py#L162-L172

lucc commented:
If I try your branch alot terminates with a message "Could not read /home/luc/.config/abook/addressbook". Sadly there are no details in the log file.

geier commented:
@lucc Thanks, I didn't look at the doctests yet at all (actually didn't notice them before)

Is that error message with python 2 or 3?

lucc commented:
All of this is with python3.

If I comment out my abook config section alot starts up fine (There is still a notice about abook stuff in the log but that file does not exist. I use a custom location for my abook). But then I can not read signed messages. The following error message is displayed in alot:
```
'GpgmeError' object has no attribute 'message'
(check the log for details)
```
 Here is the log file ("confidental" info replaced):
```
DEBUG:manager:/home/luc/.config/alot/themes
DEBUG:manager:abook defined: {'type': None, 'ignorecase': True, 'command': None, 'regexp': None, 'abook_contacts_file': '~/.abook/addressbook', 'shellcommand_external_filtering': True}
DEBUG:manager:abook defined: {'type': None, 'ignorecase': True, 'command': None, 'regexp': None, 'abook_contacts_file': '~/.abook/addressbook', 'shellcommand_external_filtering': True}
INFO:ui:setup gui in 256 colours
DEBUG:ui:fire first command
DEBUG:ui:global command string: "search tag:inbox AND NOT tag:killed"
DEBUG:__init__:mode:global got commandline "search tag:inbox AND NOT tag:killed"
DEBUG:__init__:ARGS: ['search', 'tag:inbox', 'AND', 'NOT', 'tag:killed']
DEBUG:__init__:cmd parms {'query': ['tag:inbox', 'AND', 'NOT', 'tag:killed'], 'sort': None}
DEBUG:manager:Worker process 12284 spawned
DEBUG:manager:Worker process 12284 exited successfully
DEBUG:utils:unquoted header: |somebody 1 <somebody1@example.com>|
DEBUG:utils:unquoted header: |somebody 2 <somebody2@example.com>|
DEBUG:utils:unquoted header: |somebody 3 <somebody3@example.com>|
DEBUG:walker:EMPTY PIPE
DEBUG:ui:Got key (['enter'], [10])
DEBUG:ui:cmdline: 'select'
DEBUG:ui:search command string: "select"
DEBUG:__init__:mode:search got commandline "select"
DEBUG:__init__:ARGS: ['select']
DEBUG:__init__:cmd parms {}
INFO:search:open thread view for thread:00000000000021b8: Geburstag
DEBUG:utils:unquoted header: |somebody <somebody@example.com>|
DEBUG:utils:unquoted header: |somebody else <somebody-else@example.com>|
ERROR:ui:Traceback (most recent call last):
  File "/home/luc/.local/lib/python3.5/site-packages/Twisted-15.5.0-py3.5.egg/twisted/internet/defer.py", line 317, in addCallback
    callbackKeywords=kw)
  File "/home/luc/.local/lib/python3.5/site-packages/Twisted-15.5.0-py3.5.egg/twisted/internet/defer.py", line 306, in addCallbacks
    self._runCallbacks()
  File "/home/luc/.local/lib/python3.5/site-packages/Twisted-15.5.0-py3.5.egg/twisted/internet/defer.py", line 588, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "/home/luc/.local/lib/python3.5/site-packages/alot-0.3.6-py3.5.egg/alot/ui.py", line 640, in call_apply
    return defer.maybeDeferred(cmd.apply, self)
--- <exception caught here> ---
  File "/home/luc/.local/lib/python3.5/site-packages/Twisted-15.5.0-py3.5.egg/twisted/internet/defer.py", line 150, in maybeDeferred
    result = f(*args, **kw)
  File "/home/luc/.local/lib/python3.5/site-packages/alot-0.3.6-py3.5.egg/alot/commands/search.py", line 38, in apply
    sb = buffers.ThreadBuffer(ui, self.thread)
  File "/home/luc/.local/lib/python3.5/site-packages/alot-0.3.6-py3.5.egg/alot/buffers.py", line 334, in __init__
    self.rebuild()
  File "/home/luc/.local/lib/python3.5/site-packages/alot-0.3.6-py3.5.egg/alot/buffers.py", line 362, in rebuild
    self._tree = ThreadTree(self.thread)
  File "/home/luc/.local/lib/python3.5/site-packages/alot-0.3.6-py3.5.egg/alot/widgets/thread.py", line 352, in __init__
    accumulate(msg)
  File "/home/luc/.local/lib/python3.5/site-packages/alot-0.3.6-py3.5.egg/alot/widgets/thread.py", line 331, in accumulate
    self._message[mid] = MessageTree(msg, odd)
  File "/home/luc/.local/lib/python3.5/site-packages/alot-0.3.6-py3.5.egg/alot/widgets/thread.py", line 164, in __init__
    self._maintree = SimpleTree(self._assemble_structure())
  File "/home/luc/.local/lib/python3.5/site-packages/alot-0.3.6-py3.5.egg/alot/widgets/thread.py", line 191, in _assemble_structure
    mainstruct.append((self._get_headers(), None))
  File "/home/luc/.local/lib/python3.5/site-packages/alot-0.3.6-py3.5.egg/alot/widgets/thread.py", line 254, in _get_headers
    headers)
  File "/home/luc/.local/lib/python3.5/site-packages/alot-0.3.6-py3.5.egg/alot/widgets/thread.py", line 268, in construct_header_pile
    mail = self._message.get_email()
  File "/home/luc/.local/lib/python3.5/site-packages/alot-0.3.6-py3.5.egg/alot/db/message.py", line 80, in get_email
    self._email = message_from_file(f_mail)
  File "/home/luc/.local/lib/python3.5/site-packages/alot-0.3.6-py3.5.egg/alot/db/utils.py", line 136, in message_from_file
    m.get_payload(1).get_payload())
  File "/home/luc/.local/lib/python3.5/site-packages/alot-0.3.6-py3.5.egg/alot/crypto.py", line 196, in verify_detached
    raise GPGProblem(e.message, code=e.code)
builtins.AttributeError: 'GpgmeError' object has no attribute 'message'

DEBUG:ui:Got key (['q'], [113])
DEBUG:ui:cmdline: 'bclose'
DEBUG:ui:search command string: "bclose"
DEBUG:__init__:mode:search got commandline "bclose"
DEBUG:__init__:ARGS: ['bclose']
DEBUG:__init__:cmd parms {'redraw': None, 'force': False}
INFO:globals:closing the last buffer, exiting
```

lucc commented:
Are there any news on this? Do you need more info?

pazz commented:
well, no news really: I don't know what would be the best way to merge this without breaking things for too many people. That is, how to properly test this.
Do we have a fix for the above mentioned bg for instance?
Maybe it'd be best to first add some testing suite into the main branch, so that we can more systematically test what could break when we switch to py3.. I'm open for opinions

lucc commented:
We could also split this PR. I assume that some of these things can be done in python2 with the __future__ module. And I also assume that these changes do not break anything/much so it is possible for users to test it. According to https://docs.python.org/2/library/__future__.html this could be for the print function and unicode literals.

geier commented:
I'm sorry that I didn't fix those bugs yet, I hope I find the time soon.
Integrating these changes on a commit by commit base is certainly
possible, but

a) I didn't introduce unicode_literals, I won't, and will strongly
discourage their use. I didn't believe it first, but they do cause
more problems than they fix

b) the unicode changes are probably the most drastic ones and the ones I
suspect of being at fault for the current bugs.

Quoting Lucas Hoffmann (2016-02-04 13:56:16)
> We could also split this PR. I assume that some of these things can be done in python2 with the __future__ module. And I also assume that these changes do not break anything/much so it is possible for users to test it. According to https://docs.python.org/2/library/__future__.html this could be for the print function and unicode literals.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/pazz/alot/pull/800#issuecomment-179819743

lucc commented:
@pazz to decide on a test framework maybe these can help:

- http://docs.python-guide.org/en/latest/writing/tests/
- https://wiki.python.org/moin/PythonTestingToolsTaxonomy

pazz commented:
@lucc i added a branch `0.3.8-unittests`  that introduces a unittest testsuite. Mock looks interesting for later as well..

teto commented:
I could test this if it were rebased. Looking forward to python3 support.

geier commented:
Last time I checked, there were currently some problems regarding the handling of the actual emails.

I'll try to rebase this if I find the time.

Also I still think it's probably the best way forward to start
integrating the unproblematic changes first and than work from thereon
forward.

Quoting Matthieu Coudron (2016-09-27 14:30:49)
> I could test this if it were rebased. Looking forward to python3 support.
> 
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/pazz/alot/pull/800#issuecomment-249850753

lucc commented:
#883 has some of these unproblematic fixes that make import statements "future ready".

geier commented:
I have just rebased to current master (and removed the commits not longer applicable).

This is working for me under python **2**, but a lot of functionality is still broken under python **3**, I believe all remaining issues are with unicode/byte separation.

Feel free to push onto this branch, but please leave a note so I don't accidentally force push over your changes.

pazz commented on alot/db/thread.py:
```diff
@@ -152,8 +152,8 @@ def get_authors(self):
             msgs = self.get_messages().keys()
-            msgs_with_date = filter(lambda m: m.get_date() is not None, msgs)
-            msgs_without_date = filter(lambda m: m.get_date() is None, msgs)
+            msgs_with_date = list(filter(lambda m: m.get_date() is not None, msgs))
```
why?

lucc replied:
I think because they will be sorted in the next line and that only works with lists? But then you can also use list comprehension instead of filter().

pazz commented:
some of this looks harmless to me: 1d7ac68, 70f9acd, b74e124  and I'd push these right away.

I don't understand 0ad2d8a, 6279bc8 .

I am not convinced of this new unicode_type stuff:  be029db, 552cd8f . is this the usual way to go to support both python 2 and 3?

I'd like to see some inline docs to mark where we introduce complication (i.e. in imports) in order to deal with python 3. This would make it easier later if we decide to declutter and remove support for python 2.

geier commented:
* unicode stuff: Lennart Regebro's [porting to python 3 book](http://python3porting.com/noconv.html#more-bytes-strings-and-unicode) advocates this method, as does [six](https://pythonhosted.org/six/), but it does look like the canonical type names are `text_type` and `binary_type`.
* map, lists etc: I'll need to have a look at those again.

geier commented:
inline docs are probably a good idea, but the imports are not unified at the moment anyway. Would something like https://github.com/pazz/alot/pull/800/files#diff-95a05bb59fa90905393bfa9d73ececb3R29 be good enough in your opinion?

geier commented:
note to self: crypto is getting bytes but doesn't  like it:

INFO:search:open thread view for thread:0000000000008c23: Bug#844081: Reproducer
ERROR:ui:Traceback (most recent call last):
  File "/usr/home/cg/.virtualenvs/alot/lib/python2.7/site-packages/twisted/internet/defer.py", line 317, in addCallback
    callbackKeywords=kw)
  File "/usr/home/cg/.virtualenvs/alot/lib/python2.7/site-packages/twisted/internet/defer.py", line 306, in addCallbacks
    self._runCallbacks()
  File "/usr/home/cg/.virtualenvs/alot/lib/python2.7/site-packages/twisted/internet/defer.py", line 588, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "/usr/home/cg/workspace/alot/alot/ui.py", line 639, in call_apply
    return defer.maybeDeferred(cmd.apply, self)
--- <exception caught here> ---
  File "/usr/home/cg/.virtualenvs/alot/lib/python2.7/site-packages/twisted/internet/defer.py", line 150, in maybeDeferred
    result = f(*args, **kw)
  File "/usr/home/cg/workspace/alot/alot/commands/search.py", line 38, in apply
    sb = buffers.ThreadBuffer(ui, self.thread)
  File "/usr/home/cg/workspace/alot/alot/buffers.py", line 334, in __init__
    self.rebuild()
  File "/usr/home/cg/workspace/alot/alot/buffers.py", line 362, in rebuild
    self._tree = ThreadTree(self.thread)
  File "/usr/home/cg/workspace/alot/alot/widgets/thread.py", line 356, in __init__
    accumulate(msg)
  File "/usr/home/cg/workspace/alot/alot/widgets/thread.py", line 335, in accumulate
    self._message[mid] = MessageTree(msg, odd)
  File "/usr/home/cg/workspace/alot/alot/widgets/thread.py", line 168, in __init__
    self._maintree = SimpleTree(self._assemble_structure())
  File "/usr/home/cg/workspace/alot/alot/widgets/thread.py", line 195, in _assemble_structure
    mainstruct.append((self._get_headers(), None))
  File "/usr/home/cg/workspace/alot/alot/widgets/thread.py", line 258, in _get_headers
    headers)
  File "/usr/home/cg/workspace/alot/alot/widgets/thread.py", line 272, in construct_header_pile
    mail = self._message.get_email()
  File "/usr/home/cg/workspace/alot/alot/db/message.py", line 78, in get_email
    self._email = message_from_file(f_mail)
  File "/usr/home/cg/workspace/alot/alot/db/utils.py", line 151, in message_from_file
    m.get_payload(1).get_payload())
  File "/usr/home/cg/workspace/alot/alot/crypto.py", line 204, in verify_detached
    message_data = StringIO(message)
exceptions.TypeError: initial_value must be unicode or None, not str

zenhack commented:
I want to raise the notion of using six again -- every instance of `try: ... except ImportError: ...` would go away (using six.moves instead), and all of the unicode_type stuff would be replaced with a similar construct from the library, which is more widely used and recognized. 95% of this patch so far is just re-implmenting stuff that's in six.

As far as the dependency burden is concerned, it's already a transitive dependency of *setuptools* (via python-packaging), so I don't think that should be much of a concern.

geier commented:
closed for #1055
